### PR TITLE
feat(mem_cache): add UnifiedRadixCache SWA support

### PR DESCRIPTION
**Pull request by Infra-lab-dev**
## Motivation

Implement the accepted [RFC-0035 Unified Radix Cache SWA design](https://github.com/primatrix/wiki/blob/main/docs/rfc/0035-unified-radix-cache-swa.md) for hybrid FULL+SWA models, while keeping the existing `SWARadixCache` route (Baseline) as an explicit rollback path.

Tracked by #1531.

### Routing

| Configuration | Cache implementation |
|---|---|
| Hybrid SWA + radix disabled | `SWAChunkCache` |
| Hybrid SWA + radix enabled + Unified enabled | `UnifiedRadixCache(FULL, SWA)` |
| Hybrid SWA + radix enabled + Unified disabled | `SWARadixCache` |
| Non-hybrid | Existing route, unchanged |

The new route is gated by `--enable-unified-radix-tree`. Radix-disabled routing takes precedence. Unsupported new-route combinations are rejected during `build_kv_cache()` before the tree cache is constructed, and never silently fall back: HiCache, PD/disaggregation, speculative/EAGLE, and FULL+SWA+RECURRENT.

### Ownership and lifecycle

unified-radix-cache-architecture

- Adds rank-local FULL-to-SWA page translation and grouped-free ownership.
- Adds the SWA component's window match/LRU, tombstones and healing, stable boundary locks, and internal eviction.
- Preserves locked tree-owned FULL slots during tombstone healing by transferring only the newly computed SWA mapping and releasing the temporary FULL allocation.
- Propagates complete lock receipts and the real SWA eviction boundary through request lifecycle paths.
- Keeps scheduler idle validation on the existing cache sanity check and page-aware O(1) memory accounting; `flush_cache()` has no new diagnostic gate.
- Activates explicit `UnifiedRadixCache(FULL, SWA)` construction only for the supported route.

The implementation is deliberately narrower than the full generalized design: this PR activates FULL+SWA only, retains the Baseline SWA route, and rejects recurrent/speculative/HiCache/PD combinations instead of silently falling back.

## Correctness validation

> **Evidence provenance:** The Gemma correctness/performance and MiMo-V2.5 DP-rank isolation results below were produced from PR commit `5087cb42245cc142aa8831806b4b3b2629dd588f`. The separate MiMo-V2.5-Pro AIME evaluation uses the patched source identified in its own section.

### v7x Gemma correctness gate 

Hardware/config: TPU v7x `2x2x1`, 1 host / 4 chips / 8 JAX devices; `google/gemma-4-31B-it`; TP8/DP1/attention-TP8/EP1; page size 128; greedy seed 3. Baseline and Unified used fresh servers and identical fixed inputs; the only semantic launch difference was `--enable-unified-radix-tree`.

Each of the four routes—Baseline, baseline-repeat, Unified, and Unified-repeat—completed the same 2,243-request RID set.

| Check | Result |
|---|---:|
| Baseline vs baseline-repeat | 0 mismatches across 2,243 RIDs |
| Unified vs Unified-repeat | 0 mismatches across 2,243 RIDs |
| Baseline vs Unified | 0 mismatches across 2,243 RIDs |
| Unified fixed healing cycles | 4/4 required cached-prefix behaviors observed in primary and repeat |
| Request errors | 0 |
| Offline checker | **PASS** |

### v7x MiMo-V2.5 DP-rank isolation gate

Hardware/config: TPU v7x `2x2x1`, 1 host / 4 chips / 8 JAX devices; TP8/DP2/attention-TP4/EP8; page size 256; context length 262144; SWA:FULL token ratio 0.25; round-robin DP scheduling; greedy seed 3. Baseline and Unified used fresh sequential servers and identical fixed inputs.

Each route completed 72 named requests, including 32 pressure requests per DP rank.

| Check | Result |
|---|---:|
| Requested rank vs `meta_info.dp_rank` | 72/72 matched on each route |
| Same-rank initial prefix reuse | 256 cached tokens on both ranks and routes |
| Cross-rank prefix reuse | 0 cached tokens in both directions and routes |
| Rank-0 pressure effect on rank 1 | no availability change on either route |
| Rank-1 pressure effect on rank 0 | no availability change on either route |
| Baseline vs Unified exact greedy outputs | 0 mismatches across 72 RIDs |
| Offline checker | **PASS** (0 findings) |

### MiMo-V2.5-Pro AIME accuracy evaluation

Hardware/config: TPU v7x `2x2x4`, 4 hosts / 16 chips / 32 JAX devices; `XiaomiMiMo/MiMo-V2.5-Pro`; TP32/DP4/EP32; page size 256; sliding window 128; context length 262144; SWA:FULL token ratio 0.25. Sampling: temperature 1, top-p 0.95, thinking enabled, max output tokens 131072, concurrency 16. Each route used three independent model repeats, normally running AIME25 followed by AIME26 on the same model.

| Dataset | Baseline mean (3 repeats) | Unified mean (3 repeats) | Difference |
|---|---:|---:|---:|
| AIME25 | 92.22% | 93.33% | +1.11 pp |
| AIME26 | 88.89% | 94.44%* | +5.56 pp* |

These observations show no mean accuracy regression in this sample. 
## Benchmarking and profiling 

Baseline and Unified used fresh sequential servers and identical generated workloads. Each route/workload ran 5 complete repetitions; the first 2 were discarded by the fixed protocol and the remaining 3 were aggregated. No best-run selection was used. Standard deviations below are computed across the 3 retained repetitions.

GSP means generated shared prefix. Lower TTFT is better.

| Workload | Route | p50 / p95 / p99 TTFT (ms) | Output tok/s (mean ± std) | Total tok/s (mean ± std) | Hit rate |
|---|---|---:|---:|---:|---:|
| GSP (generated shared prefix) | Baseline | 372.00 / 1427.04 / 1639.19 | 1446.36 ± 3.47 | 30125.26 ± 72.23 | 0.759644 |
| GSP (generated shared prefix) | Unified | 386.20 / 1433.81 / 1644.73 | 1443.98 ± 6.86 | 30075.70 ± 142.90 | 0.759644 |
| Random (no sharing) | Baseline | 731.94 / 1241.48 / 1257.43 | 1258.12 ± 1.88 | 11323.11 ± 16.89 | 0.011719 |
| Random (no sharing) | Unified | 736.10 / 1242.81 / 1261.38 | 1261.35 ± 13.08 | 11352.18 ± 117.74 | 0.011719 |

A/B acceptance summary:

| Workload | Unified total throughput vs Baseline | Hit-rate delta | Soft target |
|---|---:|---:|---|
| GSP (generated shared prefix) | **-0.165% (99.835% of Baseline)** | 0 | **PASS**: Unified hit rate matched Baseline |
| Random (no sharing) | **+0.257% (100.257% of Baseline)** | 0 | **PASS**: above the 95% throughput floor |

## Rollback and follow-up

- Immediate rollback: omit `--enable-unified-radix-tree` to restore `SWARadixCache`.
- Radix-disabled hybrid SWA continues to use `SWAChunkCache`.
- MiMo-V2.5 DP2 rank isolation was validated on the current PR source; broader multi-host DP coverage remains follow-up work.

## Checklist

- [x] PR body is in English.
- [x] Purpose and tracking issue are linked (`#1531`); the validated DP2 scope and remaining multi-host follow-up are explicit.
- [x] Evidence commit, hardware configuration, and fixed benchmark protocol are included.
- [x] Correctness, performance, and DP-rank isolation acceptance results are disclosed.
- [x] Rollback path and independent follow-up are explicit.

### CODE REVIEW

File: .superpowers/sdd/task-8-brief.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
+### Task 8：v7x / Gemma 4 DP1 PR gate
+
```
Comment (cjx0709) [inline .superpowers/sdd/task-8-brief.md]: maybe we can remove this file in pr list.

Comment (Infra-lab-dev) [reply .superpowers/sdd/task-8-brief.md]: I will remove it in the next commit."

File: python/sgl_jax/srt/entrypoints/http_server.py [hunk 2]
```diff
@@ -376,6 +376,25 @@ async def flush_cache():
...
+        )
+    try:
+        output = await _global_state.tokenizer_manager.get_cache_ledger()
+    except NotImplementedError:
+        logger.exception("get_cache_ledger is not implemented by the current backend")
+        return ORJSONResponse(
+            {"error": "Cache ledger is not implemented on this backend"},
+            status_code=HTTPStatus.NOT_IMPLEMENTED,
+        )
+    return dataclasses.asdict(output)
```
Comment (JamesBrianD) [inline python/sgl_jax/srt/entrypoints/http_server.py]: What's the purpose of new api

**Comment by cjx0709** [comment]
before ready for merge, I suggest that 
1. remove the file you talk or do the plan with agent like the markdown and the json file, it should never be happend in a formal pr;
2. the test file is so many and seems mostly produced by agent, do you have ever really see these test cases or just put it in the pr?
3. why swa radix cache feature changed with the http api?
Please fix these obvious issues and reviewed by youself before setting the PR to ready for reviewing, otherwise, it's a waste of everyone's time.

File: python/sgl_jax/srt/entrypoints/http_server.py (hunk 2 above)
Comment (Infra-lab-dev) [reply python/sgl_jax/srt/entrypoints/http_server.py]: API removed from this PR

File: python/sgl_jax/test/mem_cache/test_unified_swa_component.py [hunk 3]
```diff
@@ -0,0 +1,1519 @@
...
+import unittest
+from collections import defaultdict
+from types import SimpleNamespace
+
+# Must precede the first direct JAX import: this module owns a 2x2 CPU mesh.
+os.environ.setdefault("XLA_FLAGS", "--xla_force_host_platform_device_count=4")
+os.environ.setdefault("JAX_PLATFORMS", "cpu")
+
+import jax
+import jax.numpy as jnp
```
Comment (JamesBrianD) [inline python/sgl_jax/test/mem_cache/test_unified_swa_component.py:16]: Two new test file not use the CustomTestCase Class. The usability of most test cases are questionable; please check.

File: python/sgl_jax/test/mem_cache/test_unified_swa_scheduler.py [hunk 4]
```diff
@@ -0,0 +1,740 @@
+"""Scheduler-facing SWA cache contract tests."""
+
+from types import SimpleNamespace
+from unittest.mock import patch
+
+import jax
+import numpy as np
```
Comment (JamesBrianD) [inline python/sgl_jax/test/mem_cache/test_unified_swa_scheduler.py]: Same as top

File: python/sgl_jax/test/mem_cache/test_unified_swa_component.py (hunk 3 above)
Comment (Infra-lab-dev) [reply python/sgl_jax/test/mem_cache/test_unified_swa_component.py:16]: To follow the project’s test conventions, I updated the two unittest-based classes to inherit from CustomTestCase. The latest CPU CI run collected this file, and all tests passed.

File: python/sgl_jax/test/mem_cache/test_unified_swa_scheduler.py (hunk 4 above)
Comment (Infra-lab-dev) [reply python/sgl_jax/test/mem_cache/test_unified_swa_scheduler.py]: This file uses pytest because it relies on pytest parametrization and exception assertions, while CustomTestCase is a unittest base class. It is explicitly registered with runner="pytest" in run_suite.py.

**Comment by cjx0709** [comment]
@Infra-lab-dev  I think this PR mixes the core Unified SWA support with two broader changes: a cross-cache reclaim refactor and a new ownership-ledger framework.
Are these two feature nencerry for SWA unified cache? Maybe we  can narrow this PR to the pure Unified FULL+SWA path and the allocator/lock changes it strictly needs. 
If you also feel that it's unnecessary to include so many loosely related features here. Please move the scheduler ledger/snapshot logic, event counters, Chunk/legacy cache changes, and their tests into follow-up PRs, we can talk about them in another pr. Also, diagnostics should not block flush_cache() or replace the existing cache sanity check.

**Comment by Infra-lab-dev** [comment]
> @Infra-lab-dev I think this PR mixes the core Unified SWA support with two broader changes: a cross-cache reclaim refactor and a new ownership-ledger framework. Are these two feature nencerry for SWA unified cache? Maybe we can narrow this PR to the pure Unified FULL+SWA path and the allocator/lock changes it strictly needs. If you also feel that it's unnecessary to include so many loosely related features here. Please move the scheduler ledger/snapshot logic, event counters, Chunk/legacy cache changes, and their tests into follow-up PRs, we can talk about them in another pr. Also, diagnostics should not block flush_cache() or replace the existing cache sanity check.

**Comment by Infra-lab-dev** [comment]
@cjx0709 I’ve narrowed this PR to the core Unified FULL+SWA path and completed the DP-rank isolation validation on MiMo-V2.5. Please check again.

**Review by JamesBrianD — APPROVED** [review]
